lib/utmp.c

```diff
@@ -173,13 +172,15 @@ get_current_utmp(void)
 	}
 
 	if (NULL != ut) {
-		ret = XMALLOC(1, struct utmpx);
-		memcpy (ret, ut, sizeof (*ret));
+		struct utmpx *ut_copy;
+		ut_copy = XMALLOC(1, struct utmpx);
+		memcpy(ut_copy, ut, sizeof(*ut));
+		ut = ut_copy;
 	}
```

Comment:
Hmmm, this sounds like we want memdup().  That would be rewritten simpler if we had it:

```c
ut = XMEMDUP(ut, 1, struct utmpx);
```

where MEMDUP() would be

```c
#define MEMDUP(p, n, T)                                                       \
(                                                                             \
	_Generic(p, T *: (T *) memdup(p, (n * sizeof(T)) ?: 1))               \
)
```

and memdup() would be:

```c
void *
memdup(const void *p, size_t size)
{
	void  *new;

	new = malloc(size);
	if (new == NULL)
		return NULL;

	return memcpy(new, p, size);
}
```

Not saying you should do it.  It's more a reminder for myself, to check in the future.

Reply:
Plus, if this sometimes return NULL, I don't see why it uses XMALLOC().  It could pass a malloc(3) failure to the caller as NULL.